File: tt-train/sources/ttml/modules/gpt_block.cpp
// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
//
// SPDX-License-Identifier: Apache-2.0

#include "gpt_block.hpp"

#include "core/tt_tensor_utils.hpp"
#include "ops/binary_ops.hpp"
#include "ops/unary_ops.hpp"

namespace ttml::modules {

GPTMLP::GPTMLP(uint32_t embedding_size, float dropout_prob) {
    fc1 = std::make_shared<LinearLayer>(embedding_size, embedding_size * 4);
    fc2 = std::make_shared<LinearLayer>(embedding_size * 4, embedding_size);
    dropout = std::make_shared<DropoutLayer>(dropout_prob);

    create_name("gpt_mlp");
    register_module(fc1, "fc1");
    register_module(fc2, "fc2");
    register_module(dropout, "dropout");
}

autograd::TensorPtr GPTMLP::operator()(const autograd::TensorPtr& input) {
    auto x = (*fc1)(input);
    x = ops::gelu(x);
    x = (*fc2)(x);
    x = (*dropout)(x);
    return x;
}

GPTBlock::GPTBlock(uint32_t embedding_size, uint32_t num_heads, float dropout_prob, bool use_composite_layernorm) {
    mlp = std::make_shared<GPTMLP>(embedding_size, dropout_prob);
    ln1 = std::make_shared<LayerNormLayer>(embedding_size, use_composite_layernorm);
    ln2 = std::make_shared<LayerNormLayer>(embedding_size, use_composite_layernorm);
    attention = std::make_shared<MultiHeadAttention>(embedding_size, num_heads, dropout_prob);

    create_name("gpt_block");
    register_module(mlp, "mlp");
    register_module(ln1, "ln1");
    register_module(ln2, "ln2");
    register_module(attention, "attention");
}

autograd::TensorPtr GPTBlock::operator()(const autograd::TensorPtr& input, const autograd::TensorPtr& mask) {
    auto residual = input;
    auto x = (*ln1)(input);
    x = (*attention)(x, mask);
    x = ops::add(x, residual);

    residual = x;
    x = (*ln2)(x);
    x = (*mlp)(x);
    x = ops::add(x, residual);

    return x;
}

}  // namespace ttml::modules

File: tt-train/sources/ttml/ttml/models/nanogpt/gpt_block.py
# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0

"""GPT transformer block for NanoGPT."""

from __future__ import annotations

from typing import Optional

import ttml
from ttml.modules import AbstractModuleBase, Parameter

from .gpt_mlp import GPTMLP
from .multi_head_attention import MultiHeadAttention


class LayerNorm(AbstractModuleBase):
    """Layer normalization module with gamma and beta parameters."""

    def __init__(
        self,
        embedding_dim: int,
        bias: bool = True,
        use_composite_layernorm: bool = False,
    ) -> None:
        super().__init__()

        self.embedding_dim = embedding_dim
        self.use_composite_layernorm = use_composite_layernorm

        ln_shape = (1, 1, 1, embedding_dim)
        self.gamma = Parameter(ttml.init.ones()(ln_shape))

        if bias:
            self.beta = Parameter(ttml.init.zeros()(ln_shape))
        else:
            self.beta = None

    def forward(self, x: ttml.autograd.Tensor) -> ttml.autograd.Tensor:
        """Forward pass of layer norm.

        Args:
            x: Input tensor

        Returns:
            Normalized output tensor
        """

        if self.use_composite_layernorm:
            layernorm_op = ttml.ops.layernorm.composite_layernorm
        else:
            layernorm_op = ttml.ops.layernorm.layernorm

        return layernorm_op(x, self.gamma.tensor, self.beta.tensor if self.beta else None)


class GPTBlock(AbstractModuleBase):
    """GPT transformer block with attention and MLP."""

    def __init__(
        self,
        embedding_dim: int,
        num_heads: int,
        dropout: float = 0.0,
        bias: bool = True,
        use_composite_layernorm: bool = False,
    ) -> None:
        super().__init__()

        self.embedding_dim = embedding_dim

        # Layer norms
        self.ln1 = LayerNorm(embedding_dim, bias, use_composite_layernorm)
        self.ln2 = LayerNorm(embedding_dim, bias, use_composite_layernorm)

        # Attention and MLP
        self.attention = MultiHeadAttention(
            embedding_dim,
            num_heads,
            dropout,
        )
        self.mlp = GPTMLP(
            embedding_dim,
            dropout,
        )

    # train() and eval() are inherited from AbstractModuleBase
    # They automatically propagate RunMode to all registered submodules

    def forward(self, x: ttml.autograd.Tensor, mask: Optional[ttml.autograd.Tensor] = None) -> ttml.autograd.Tensor:
        """Forward pass of GPT block.

        Args:
            x: Input tensor
            mask: Optional attention mask

        Returns:
            Output tensor after transformer block
        """
        # Pre-norm attention with residual (matching C++)
        # residual = input; x = (*ln1)(input); x = (*attention)(x, mask); x = ops::add(x, residual)
        residual = x
        x = self.ln1(x)
        x = self.attention(x, mask)
        x = ttml.ops.binary.add(x, residual)

        # Pre-norm MLP with residual (matching C++)
        # residual = x; x = (*ln2)(x); x = (*mlp)(x); x = ops::add(x, residual)
        residual = x
        x = self.ln2(x)
        x = self.mlp(x)
        x = ttml.ops.binary.add(x, residual)

        return x
